runtime/commonMain/src/kotlinx/serialization/json/JsonParametricSerializer.kt

```diff
@@ -0,0 +1,96 @@
+ * (often when interacting with external API) such property is not present in the input
+ * and one should guess type using other evidence, e.g. presence of some keys in the object.
+ * This class offers a skeleton implementation of such guessing strategy. Please note that
+ * since JSON content is represented by [JsonElement] class and could be read only with [JsonInput] decoder,
+ * this class works only with [Json] format.
+ *
+ * Deserialization happens in two stages: first, a value from the JSON stream is read
+ * to a [JsonElement]. Then, [selectSerializer] function is called to determine which serializer should be used.
+ * The returned serializer is used to deserialize [JsonElement] back to Kotlin object.
+ *
+ * One could also serialize a value with this serializer. In that case, class discriminator property would not
+ * be added to JSON stream, i.e., deserializing a class from the string and serializing it back yields the original string
+ * However, to determine a serializer, a standard polymorphic mechanism represented by [SerialModule] is used.
+ * For convenience, [serialize] method can lookup default serializer, but it is recommended to follow
+ * standard procedure with [registering][SerializersModuleBuilder.polymorphic].
+ *
+ * Usage example:
+ * ```
+ * interface Payment {
+ *     val amount: String
+ * }
+ *
+ * @Serializable
+ * data class SuccessfulPayment(override val amount: String, val date: String) : Payment
+ *
+ * @Serializable
+ * data class RefundedPayment(override val amount: String, val date: String, val reason: String) : Payment
+ *
+ * object PaymentSerializer : JsonParametricSerializer<Payment>(Payment::class) {
+ *     override fun selectSerializer(element: JsonElement): KSerializer<out Payment> = when {
+ *         "reason" in element -> RefundedPayment.serializer()
+ *         else -> SuccessfulPayment.serializer()
+ *     }
+ * }
+ *
+ * // Now both statements will yield different subclasses of Payment:
+ *
+ * Json.parse(PaymentSerializer, """{"amount":"1.0","date":"03.02.2020"}""")
+ * Json.parse(PaymentSerializer, """{"amount":"2.0","date":"03.02.2020","reason":"complaint"}""")
+ * ```
+ *
+ * @param T A root class for all classes that could be possibly encountered during serialization and deserialization.
+ * @param baseClass A class token for [T].
+ *
+ */
+public abstract class JsonParametricSerializer<T : Any>(private val baseClass: KClass<T>) : KSerializer<T> {
+    /**
+     * A descriptor for this set of parametric serializers.
+     * By default, it uses the name composed of [baseClass] simple name,
+     * kind is set to [PolymorphicKind.SEALED] and contains 0 elements.
+     *
+     * However, this descriptor can be overridden to achieve better representation of custom transformed JSON shape
+     * for schema generating/introspection purposes.
+     */
+    override val descriptor: SerialDescriptor =
```

Comment:
Great idea with a pre-built descriptor.
For the sake of consistency, let's change its type to `CONTEXTUAL` because it's not really about sealed classes

Reply:
After #682, because `CONTEXTUAL` is not yet in dev )